File: 04-Queues/20250727-beginner/go/solution.go
package main

// TODO: Implement a Queue data structure with the following operations:
// - Enqueue(item): Add an item to the back of the queue
// - Dequeue(): Remove and return the item from the front of the queue
// - Front(): Return the front item without removing it
// - IsEmpty(): Check if the queue is empty
// - Size(): Return the number of items in the queue

// TODO: Implement your Queue structure here
type Queue struct {
	// TODO: Define your queue fields
	data []any
}

// TODO: Implement NewQueue constructor
func NewQueue() *Queue {
	// TODO: Initialize and return a new queue
	queue := &Queue{
		data: nil,
	}
	return queue
}

// TODO: Implement Enqueue method
func (q *Queue) Enqueue(item any) {
	// TODO: Add item to the back of the queue
	if q.data == nil {
		return
	}
	q.data = append(q.data, item)
}

// TODO: Implement Dequeue method
func (q *Queue) Dequeue() any {
	// TODO: Remove and return the front item
	// Return nil if queue is empty
	if q.Size() == 0 {
		return nil
	}
	return nil
}

// TODO: Implement Front method
func (q *Queue) Front() any {
	// TODO: Return the front item without removing it
	// Return nil if queue is empty
	if q.Size() == 0 {
		return nil
	}
	return nil
}

// TODO: Implement IsEmpty method
func (q *Queue) IsEmpty() bool {
	// TODO: Return true if queue is empty, false otherwise
	return q.Size() == 0
}

// TODO: Implement Size method
func (q *Queue) Size() int {
	// TODO: Return the number of items in the queue
	return len(q.data)
}

File: 04-Queues/20250727-beginner/ts/solution.ts
// TODO: Implement a Queue data structure with the following operations:
// - enqueue(item): Add an item to the back of the queue
// - dequeue(): Remove and return the item from the front of the queue
// - front(): Return the front item without removing it
// - isEmpty(): Check if the queue is empty
// - size(): Return the number of items in the queue

export class Queue<T> {
  // TODO: Define your queue properties

  constructor() {
    // TODO: Initialize your queue
  }

  // TODO: Implement enqueue method
  enqueue(item: T): void {
    // TODO: Add item to the back of the queue
  }

  // TODO: Implement dequeue method
  dequeue(): T | null {
    // TODO: Remove and return the front item
    // Return null if queue is empty
    return null;
  }

  // TODO: Implement front method
  front(): T | null {
    // TODO: Return the front item without removing it
    // Return null if queue is empty
    return null;
  }

  // TODO: Implement isEmpty method
  isEmpty(): boolean {
    // TODO: Return true if queue is empty, false otherwise
    return false;
  }

  // TODO: Implement size method
  size(): number {
    // TODO: Return the number of items in the queue
    return 0;
  }
}